Support SNARK seals in Bonsai relay and enable CI for Solidity code via Foundry

nategraf opened the pull request:
This PR:
* Adds support for supplying SNARKs to authenticate callbacks through the `BonsaiRelay` contract, thus enabling the fully-verifying Bonsai callback flow for use in local deployment and testnet environments such as Sepolia.
* Updates the Governance showcase to be based on the latest version of the Foundry template, including supporting the `run` and `upload` subcommands in the local relay CLI.
* Adds GitHub actions workflows for testing and linting Solidity code with Foundry. It includes actions to trigger `forge fmt` and `forge test` in `bonsai/ethereum` (the Solidity library for Bonsai) and `bonsai/examples/governance` (Governance showcase).

This PR is a dependency of https://github.com/risc0/bonsai-foundry-template/pull/29

Resolves https://github.com/risc0/risc0/issues/722

nategraf commented:
@imdaisyadams one unresolved issue is getting the end-to-end test flow setup hitting a Bonsai endpoint. I'm not 100% sure how to handle this from a standpoint of maintaining reliability in our CI workflows (i.e. maybe we don't actually want to include the end-to-end test at all?)

flaub commented on bonsai/examples/governance/Cargo.toml:
```diff
@@ -1,12 +1,19 @@
 
+[workspace.dependencies]
+# Outside of the risc0 monorepo, use the following dependency definitions.
+# risc0-build = { git = "https://github.com/risc0/risc0", branch = "release-0.16" }
+# risc0-zkvm = { git = "https://github.com/risc0/risc0", branch = "release-0.16", default-features = false }
+# bonsai-sdk = { git = "https://github.com/risc0/risc0", branch = "release-0.16" }
```
It's legal to put both `path = ` and `version =`. This allows in-tree and out-tree building to work.

nategraf commented:
@morgangibbs87 or @imdaisyadams, do you have any idea what might be going on here
https://github.com/risc0/risc0/actions/runs/5593389447/jobs/10226980254?pr=712

nategraf commented:
> @nategraf - The error seems to show that `DEP_RISC0_SYS_METAL_KERNELS_ZKP` is not defined somewhere.
> 
> ```
> --- stderr
>     thread 'main' panicked at 'CARGO_FEATURE_METAL is defined, but DEP_RISC0_SYS_METAL_KERNELS_ZKP is not: NotPresent', /Users/risc0/actions-runner/_work/risc0/risc0/risc0/zkp/build.rs:26:14
> ```

Ya, so what's unclear to me at the moment is how this PR would have triggered this issue. I will investigate.

flaub commented on risc0/zkp/src/core/hash/poseidon/mod.rs:
```diff
@@ -15,6 +15,8 @@
 //! An implementation of Poseidon targeting the Baby Bear field with a security
 //! of 128 bits.
 
+extern crate alloc;
```
Why do we need these? If we do need it, it should be sufficient to specify it once at the top-level in `lib.rs`

nategraf replied:
Currently `cargo clippy` thinks the crate is unsued when specified in `mod.rs`, so on first pass I made this change, but looking at it again it makes more sense just to turn off the warning.

flaub commented on bonsai/examples/governance/methods/guest/Cargo.toml:
```diff
@@ -8,10 +8,11 @@ name = "finalize_votes"
 path = "src/bin/finalize_votes.rs"
 
 [dependencies]
+risc0-zkvm = { version = "0.16", path = "../../../../../risc0/zkvm", default-features = false, features = ["std"] }
```
Please sort deps

nategraf replied:
Done in cf1c802c4328f19f3d6d32961dd985c1d6acb463

flaub commented on bonsai/examples/governance/methods/Cargo.toml:
```diff
@@ -1,17 +1,22 @@
 [package]
-name = "bonsai-starter-methods"
+name = "methods"
```
I think we should still prefix this, maybe `governance-methods`

nategraf replied:
Done in cf1c802c4328f19f3d6d32961dd985c1d6acb463

flaub commented on .github/workflows/main.yml:
```diff
@@ -77,8 +78,7 @@ jobs:
       - run: cargo test -F $FEATURE --tests -- --ignored
       - run: cargo test -F $FEATURE --manifest-path examples/Cargo.toml
-      - run: cargo test --manifest-path bonsai/examples/governance/Cargo.toml
-        if: matrix.device == 'cpu'
```
Was this intentional? This is probably what's causing all the build failures. You need to specify `-F $FEATURE` if you want this to work on GPUs. But I also wonder what we gain by running this test on GPU.

flaub replied:
I think we should also considering making the workspace `bonsai/examples` rather than `bonsai/examples/governance`

nategraf replied:
> I think we should also considering making the workspace bonsai/examples rather than bonsai/examples/governance

The main goal of the showcase is to demonstrate what a more-or-less full project looks like. If pull the workspace declarations out into `bonsai/examples` it makes it harder to copy this as a self-contained starting point/example. Maybe we can consider this again when we add a second Bonsai example? By then we will hopefully have also extracted the `relay` crate, which is not actually specific to this example.

flaub replied:
I'm starting to think maybe we can extend the `risczero` tool to turn examples into standalone projects.

imdaisyadams commented on .github/workflows/foundry.yml:
```diff
@@ -0,0 +1,57 @@
+        uses: risc0/foundry-toolchain@2fe7e70b520f62368a0e3c464f997df07ede420f
+
+      - name: check solidity code formatting in bonsai/ethereum
+        run: forge fmt --check
```
could we make this one line? `forget fmt --check bonsai/ethereum`. same with L39?

nategraf replied:
It seems this actually yield different results. It seems in general `forge` expects the project root to be the current directory.

nategraf commented:
This PR is (finally) ready for review 🎉 

I did add one todo to the README. I think this should be addressed in the future as we make further improvements to the relay. https://github.com/risc0/risc0/pull/712/files#diff-2af8b9be5e649349cbac2e06a7f6dcd14c3903750e7ca741ea88640a68f0ac65R45

flaub commented on bonsai/examples/governance/methods/guest/Cargo.toml:
```diff
@@ -17,9 +17,7 @@ k256 = { version = "=0.13.1", features = [
-risc0-zkvm = { path = "../../../../../risc0/zkvm", default-features = false, features = [
-  "std",
-] }
+risc0-zkvm = { version = "0.16", path = "../../../../../risc0/zkvm", default-features = false, features = ["std"] }
```
`version =` is only needed if we plan to publish the crate

nategraf replied:
Removed in https://github.com/risc0/risc0/pull/712/commits/dc7473d831779b05ac6c657ce7ba399c3d030a0b

nategraf commented:
I've updated the control ID values to reflect that latest version of the "identity p254" circuit, and updated the tests for `RiscZeroGroth16Verifier.sol` to use a seal produced by the dev deployment of Bonsai proving service. End to end tests against the dev deployment of the Bonsai proving service now pass as well.

tzerrell commented on bonsai/ethereum/contracts/IBonsaiRelay.sol:
```diff
@@ -38,4 +62,22 @@ interface IBonsaiRelay {
         uint64 gasLimit
     ) external;
+
+    /// @notice Determines if the given athorization is valid for the image ID and journal.
+    /// @dev A (imageId, jounral) pair should be valid, and the respective callback authorized, if
```
Typos
```suggestion
    /// @notice Determines if the given authorization is valid for the image ID and journal.
    /// @dev A (imageId, journal) pair should be valid, and the respective callback authorized, if
```
